gobblin-modules/gobblin-kafka-09/src/main/java/gobblin/source/extractor/extract/kafka/KafkaSimpleStreamingExtractor.java

```diff
@@ -0,0 +1,227 @@
+
+import com.google.common.base.Preconditions;
+import com.google.gson.Gson;
+import com.google.gson.JsonElement;
+
+
+import gobblin.configuration.WorkUnitState;
+import gobblin.configuration.State;
+import gobblin.metrics.kafka.KafkaSchemaRegistry;
+import gobblin.metrics.kafka.SchemaRegistryException;
+import gobblin.source.extractor.DataRecordException;
+import gobblin.source.extractor.Watermark;
+import gobblin.source.extractor.WatermarkSerializerHelper;
+import gobblin.source.extractor.extract.EventBasedExtractor;
+import gobblin.source.extractor.StreamingExtractor;
+import gobblin.source.extractor.RecordEnvelope;
+import gobblin.source.extractor.CheckpointableWatermark;
+import gobblin.source.extractor.ComparableWatermark;
+import gobblin.source.extractor.extract.LongWatermark;
+import gobblin.metrics.Tag;
+import gobblin.writer.WatermarkStorage;
+
+
+/**
+ * An implementation of {@link StreamingExtractor} from which reads from Kafka and returns records as an array of bytes.
+ *
+ * @author Shrikanth Shankar
+ *
+ *
+ */
+public class KafkaSimpleStreamingExtractor<S, D> extends EventBasedExtractor<S, RecordEnvelope<D>> implements StreamingExtractor<S, D>, WatermarkStorage {
+
+  public static class KafkaWatermark implements CheckpointableWatermark {
+    TopicPartition _topicPartition;
+    LongWatermark _lwm;
+
+    @VisibleForTesting
+    public KafkaWatermark(TopicPartition topicPartition, LongWatermark lwm) {
+      _topicPartition = topicPartition;
+      _lwm = lwm;
+    }
+
+    @Override
+    public String getSource() {
+      return _topicPartition.toString();
+    }
+
+    @Override
+    public ComparableWatermark getWatermark() {
+      return _lwm;
+    }
+
+    @Override
+    public short calculatePercentCompletion(Watermark lowWatermark, Watermark highWatermark) {
+      return 0;
+    }
+
+    @Override
+    public JsonElement toJson() {
+      return WatermarkSerializerHelper.convertWatermarkToJson(this);
+    }
+
+    @Override
+    public int compareTo(CheckpointableWatermark o) {
+      Preconditions.checkArgument(o instanceof KafkaWatermark);
+      KafkaWatermark ko = (KafkaWatermark)o;
+      Preconditions.checkArgument(_topicPartition.equals(ko._topicPartition));
+      return _lwm.compareTo(ko._lwm);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+      if (obj == null)
+        return false;
+      if (!(obj instanceof KafkaWatermark))
+          return false;
+      return this.compareTo((CheckpointableWatermark)obj) == 0;
+    }
+
+    @Override
+    public int hashCode() {
+      final int prime = 31;
+      return _topicPartition.hashCode() * prime + _lwm.hashCode();
+    }
+
+    public TopicPartition getTopicPartition() {return _topicPartition;};
+
+    public LongWatermark getLwm() {return  _lwm;};
+  }
+
+
+  private final Consumer<S, D> _consumer;
+  private final TopicPartition _partition;
+  private Iterator<ConsumerRecord<S, D>> _records;
+  long _rowCount = 0;
```

Comment:
If you want the whole class to be thread-safe, then this needs to be an AtomicLong. 
readRecords and getExpectedRecordCount *could* be called from different threads.